README.md
# Gen Tags

用于快速生成和推送 tag。

![view](src/assets/view.jpg)

## Features

- 按规则生成 tag
- 支持前缀和后缀
- 将生成的 tag 存在 package.json 中

> 前缀通过 package.json 中的 `tagPrefix` 字段配置，如： `tagPrefix: ["v-"]`。

## Requirements

- vscode.git 扩展

## Known Issues

暂无

## Release Notes

### 1.0.0

Initial release of ...

## For more information

* [VSCode Marketplace](https://marketplace.visualstudio.com/items?itemName=zhaoqing.gen-tags)


CHANGELOG.md
# Change Log

All notable changes to the "gen-tags" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/git.ts
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

import { Uri, Event, Disposable, ProviderResult, Command, CancellationToken, ThemeIcon } from 'vscode';
export { ProviderResult } from 'vscode';

export interface Git {
	readonly path: string;
}

export interface InputBox {
	value: string;
}

export const enum ForcePushMode {
	Force,
	ForceWithLease,
	ForceWithLeaseIfIncludes,
}

export const enum RefType {
	Head,
	RemoteHead,
	Tag
}

export interface Ref {
	readonly type: RefType;
	readonly name?: string;
	readonly commit?: string;
	readonly remote?: string;
}

export interface UpstreamRef {
	readonly remote: string;
	readonly name: string;
	readonly commit?: string;
}

export interface Branch extends Ref {
	readonly upstream?: UpstreamRef;
	readonly ahead?: number;
	readonly behind?: number;
}

export interface CommitShortStat {
	readonly files: number;
	readonly insertions: number;
	readonly deletions: number;
}

export interface Commit {
	readonly hash: string;
	readonly message: string;
	readonly parents: string[];
	readonly authorDate?: Date;
	readonly authorName?: string;
	readonly authorEmail?: string;
	readonly commitDate?: Date;
	readonly shortStat?: CommitShortStat;
}

export interface Submodule {
	readonly name: string;
	readonly path: string;
	readonly url: string;
}

export interface Remote {
	readonly name: string;
	readonly fetchUrl?: string;
	readonly pushUrl?: string;
	readonly isReadOnly: boolean;
}

export const enum Status {
	INDEX_MODIFIED,
	INDEX_ADDED,
	INDEX_DELETED,
	INDEX_RENAMED,
	INDEX_COPIED,

	MODIFIED,
	DELETED,
	UNTRACKED,
	IGNORED,
	INTENT_TO_ADD,
	INTENT_TO_RENAME,
	TYPE_CHANGED,

	ADDED_BY_US,
	ADDED_BY_THEM,
	DELETED_BY_US,
	DELETED_BY_THEM,
	BOTH_ADDED,
	BOTH_DELETED,
	BOTH_MODIFIED
}

export interface Change {

	/**
	 * Returns either `originalUri` or `renameUri`, depending
	 * on whether this change is a rename change. When
	 * in doubt always use `uri` over the other two alternatives.
	 */
	readonly uri: Uri;
	readonly originalUri: Uri;
	readonly renameUri: Uri | undefined;
	readonly status: Status;
}

export interface RepositoryState {
	readonly HEAD: Branch | undefined;
	readonly refs: Ref[];
	readonly remotes: Remote[];
	readonly submodules: Submodule[];
	readonly rebaseCommit: Commit | undefined;

	readonly mergeChanges: Change[];
	readonly indexChanges: Change[];
	readonly workingTreeChanges: Change[];

	readonly onDidChange: Event<void>;
}

export interface RepositoryUIState {
	readonly selected: boolean;
	readonly onDidChange: Event<void>;
}

/**
 * Log options.
 */
export interface LogOptions {
	/** Max number of log entries to retrieve. If not specified, the default is 32. */
	readonly maxEntries?: number;
	readonly path?: string;
	/** A commit range, such as "0a47c67f0fb52dd11562af48658bc1dff1d75a38..0bb4bdea78e1db44d728fd6894720071e303304f" */
	readonly range?: string;
	readonly reverse?: boolean;
	readonly sortByAuthorDate?: boolean;
	readonly shortStats?: boolean;
}

export interface CommitOptions {
	all?: boolean | 'tracked';
	amend?: boolean;
	signoff?: boolean;
	signCommit?: boolean;
	empty?: boolean;
	noVerify?: boolean;
	requireUserConfig?: boolean;
	useEditor?: boolean;
	verbose?: boolean;
	/**
	 * string    - execute the specified command after the commit operation
	 * undefined - execute the command specified in git.postCommitCommand
	 *             after the commit operation
	 * null      - do not execute any command after the commit operation
	 */
	postCommitCommand?: string | null;
}

export interface FetchOptions {
	remote?: string;
	ref?: string;
	all?: boolean;
	prune?: boolean;
	depth?: number;
}

export interface InitOptions {
	defaultBranch?: string;
}

export interface RefQuery {
	readonly contains?: string;
	readonly count?: number;
	readonly pattern?: string;
	readonly sort?: 'alphabetically' | 'committerdate';
}

export interface BranchQuery extends RefQuery {
	readonly remote?: boolean;
}

export interface Repository {

	readonly rootUri: Uri;
	readonly inputBox: InputBox;
	readonly state: RepositoryState;
	readonly ui: RepositoryUIState;

	getConfigs(): Promise<{ key: string; value: string; }[]>;
	getConfig(key: string): Promise<string>;
	setConfig(key: string, value: string): Promise<string>;
	getGlobalConfig(key: string): Promise<string>;

	getObjectDetails(treeish: string, path: string): Promise<{ mode: string, object: string, size: number }>;
	detectObjectType(object: string): Promise<{ mimetype: string, encoding?: string }>;
	buffer(ref: string, path: string): Promise<Buffer>;
	show(ref: string, path: string): Promise<string>;
	getCommit(ref: string): Promise<Commit>;

	add(paths: string[]): Promise<void>;
	revert(paths: string[]): Promise<void>;
	clean(paths: string[]): Promise<void>;

	apply(patch: string, reverse?: boolean): Promise<void>;
	diff(cached?: boolean): Promise<string>;
	diffWithHEAD(): Promise<Change[]>;
	diffWithHEAD(path: string): Promise<string>;
	diffWith(ref: string): Promise<Change[]>;
	diffWith(ref: string, path: string): Promise<string>;
	diffIndexWithHEAD(): Promise<Change[]>;
	diffIndexWithHEAD(path: string): Promise<string>;
	diffIndexWith(ref: string): Promise<Change[]>;
	diffIndexWith(ref: string, path: string): Promise<string>;
	diffBlobs(object1: string, object2: string): Promise<string>;
	diffBetween(ref1: string, ref2: string): Promise<Change[]>;
	diffBetween(ref1: string, ref2: string, path: string): Promise<string>;

	hashObject(data: string): Promise<string>;

	createBranch(name: string, checkout: boolean, ref?: string): Promise<void>;
	deleteBranch(name: string, force?: boolean): Promise<void>;
	getBranch(name: string): Promise<Branch>;
	getBranches(query: BranchQuery, cancellationToken?: CancellationToken): Promise<Ref[]>;
	getBranchBase(name: string): Promise<Branch | undefined>;
	setBranchUpstream(name: string, upstream: string): Promise<void>;

	getRefs(query: RefQuery, cancellationToken?: CancellationToken): Promise<Ref[]>;

	getMergeBase(ref1: string, ref2: string): Promise<string | undefined>;

	tag(name: string, upstream: string): Promise<void>;
	deleteTag(name: string): Promise<void>;

	status(): Promise<void>;
	checkout(treeish: string): Promise<void>;

	addRemote(name: string, url: string): Promise<void>;
	removeRemote(name: string): Promise<void>;
	renameRemote(name: string, newName: string): Promise<void>;

	fetch(options?: FetchOptions): Promise<void>;
	fetch(remote?: string, ref?: string, depth?: number): Promise<void>;
	pull(unshallow?: boolean): Promise<void>;
	push(remoteName?: string, branchName?: string, setUpstream?: boolean, force?: ForcePushMode): Promise<void>;

	blame(path: string): Promise<string>;
	log(options?: LogOptions): Promise<Commit[]>;

	commit(message: string, opts?: CommitOptions): Promise<void>;
}

export interface RemoteSource {
	readonly name: string;
	readonly description?: string;
	readonly url: string | string[];
}

export interface RemoteSourceProvider {
	readonly name: string;
	readonly icon?: string; // codicon name
	readonly supportsQuery?: boolean;
	getRemoteSources(query?: string): ProviderResult<RemoteSource[]>;
	getBranches?(url: string): ProviderResult<string[]>;
	publishRepository?(repository: Repository): Promise<void>;
}

export interface RemoteSourcePublisher {
	readonly name: string;
	readonly icon?: string; // codicon name
	publishRepository(repository: Repository): Promise<void>;
}

export interface Credentials {
	readonly username: string;
	readonly password: string;
}

export interface CredentialsProvider {
	getCredentials(host: Uri): ProviderResult<Credentials>;
}

export interface PostCommitCommandsProvider {
	getCommands(repository: Repository): Command[];
}

export interface PushErrorHandler {
	handlePushError(repository: Repository, remote: Remote, refspec: string, error: Error & { gitErrorCode: GitErrorCodes }): Promise<boolean>;
}

export interface BranchProtection {
	readonly remote: string;
	readonly rules: BranchProtectionRule[];
}

export interface BranchProtectionRule {
	readonly include?: string[];
	readonly exclude?: string[];
}

export interface BranchProtectionProvider {
	onDidChangeBranchProtection: Event<Uri>;
	provideBranchProtection(): BranchProtection[];
}

export interface CommitMessageProvider {
	readonly title: string;
	readonly icon?: Uri | { light: Uri, dark: Uri } | ThemeIcon;
	provideCommitMessage(repository: Repository, changes: string[], cancellationToken?: CancellationToken): Promise<string | undefined>;
}

export type APIState = 'uninitialized' | 'initialized';

export interface PublishEvent {
	repository: Repository;
	branch?: string;
}

export interface API {
	readonly state: APIState;
	readonly onDidChangeState: Event<APIState>;
	readonly onDidPublish: Event<PublishEvent>;
	readonly git: Git;
	readonly repositories: Repository[];
	readonly onDidOpenRepository: Event<Repository>;
	readonly onDidCloseRepository: Event<Repository>;

	toGitUri(uri: Uri, ref: string): Uri;
	getRepository(uri: Uri): Repository | null;
	init(root: Uri, options?: InitOptions): Promise<Repository | null>;
	openRepository(root: Uri): Promise<Repository | null>

	registerRemoteSourcePublisher(publisher: RemoteSourcePublisher): Disposable;
	registerRemoteSourceProvider(provider: RemoteSourceProvider): Disposable;
	registerCredentialsProvider(provider: CredentialsProvider): Disposable;
	registerPostCommitCommandsProvider(provider: PostCommitCommandsProvider): Disposable;
	registerPushErrorHandler(handler: PushErrorHandler): Disposable;
	registerBranchProtectionProvider(root: Uri, provider: BranchProtectionProvider): Disposable;
	registerCommitMessageProvider(provider: CommitMessageProvider): Disposable;
}

export interface GitExtension {

	readonly enabled: boolean;
	readonly onDidChangeEnablement: Event<boolean>;

	/**
	 * Returns a specific API version.
	 *
	 * Throws error if git extension is disabled. You can listen to the
	 * [GitExtension.onDidChangeEnablement](#GitExtension.onDidChangeEnablement) event
	 * to know when the extension becomes enabled/disabled.
	 *
	 * @param version Version number.
	 * @returns API instance
	 */
	getAPI(version: 1): API;
}

export const enum GitErrorCodes {
	BadConfigFile = 'BadConfigFile',
	AuthenticationFailed = 'AuthenticationFailed',
	NoUserNameConfigured = 'NoUserNameConfigured',
	NoUserEmailConfigured = 'NoUserEmailConfigured',
	NoRemoteRepositorySpecified = 'NoRemoteRepositorySpecified',
	NotAGitRepository = 'NotAGitRepository',
	NotAtRepositoryRoot = 'NotAtRepositoryRoot',
	Conflict = 'Conflict',
	StashConflict = 'StashConflict',
	UnmergedChanges = 'UnmergedChanges',
	PushRejected = 'PushRejected',
	ForcePushWithLeaseRejected = 'ForcePushWithLeaseRejected',
	ForcePushWithLeaseIfIncludesRejected = 'ForcePushWithLeaseIfIncludesRejected',
	RemoteConnectionError = 'RemoteConnectionError',
	DirtyWorkTree = 'DirtyWorkTree',
	CantOpenResource = 'CantOpenResource',
	GitNotFound = 'GitNotFound',
	CantCreatePipe = 'CantCreatePipe',
	PermissionDenied = 'PermissionDenied',
	CantAccessRemote = 'CantAccessRemote',
	RepositoryNotFound = 'RepositoryNotFound',
	RepositoryIsLocked = 'RepositoryIsLocked',
	BranchNotFullyMerged = 'BranchNotFullyMerged',
	NoRemoteReference = 'NoRemoteReference',
	InvalidBranchName = 'InvalidBranchName',
	BranchAlreadyExists = 'BranchAlreadyExists',
	NoLocalChanges = 'NoLocalChanges',
	NoStashFound = 'NoStashFound',
	LocalChangesOverwritten = 'LocalChangesOverwritten',
	NoUpstreamBranch = 'NoUpstreamBranch',
	IsInSubmodule = 'IsInSubmodule',
	WrongCase = 'WrongCase',
	CantLockRef = 'CantLockRef',
	CantRebaseMultipleBranches = 'CantRebaseMultipleBranches',
	PatchDoesNotApply = 'PatchDoesNotApply',
	NoPathFound = 'NoPathFound',
	UnknownPath = 'UnknownPath',
	EmptyCommitMessage = 'EmptyCommitMessage',
	BranchFastForwardRejected = 'BranchFastForwardRejected',
	BranchNotYetBorn = 'BranchNotYetBorn',
	TagConflict = 'TagConflict'
}

src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from 'vscode';
import TagProvider from './provider';

// This method is called when your extension is activated
// Your extension is activated the very first time the command is executed
export function activate(context: vscode.ExtensionContext) {

	const provider =  new TagProvider(context.extensionUri);
	context.subscriptions.push(
		vscode.window.registerWebviewViewProvider(TagProvider.viewType, provider, {
			webviewOptions: {
				retainContextWhenHidden: true
			}
		})
	);

	const register = (name: string, cb: () => void) => {
		return vscode.commands.registerCommand(name, cb);
	};

	context.subscriptions.push(
		register('gentags.refresh', () => provider.init()),
	);
	provider.init();
}

// This method is called when your extension is deactivated
export function deactivate() {}


webpack.config.js
//@ts-check

'use strict';

const path = require('path');

//@ts-check
/** @typedef {import('webpack').Configuration} WebpackConfig **/

/** @type WebpackConfig */
const extensionConfig = {
  target: 'node', // VS Code extensions run in a Node.js-context 📖 -> https://webpack.js.org/configuration/node/
	mode: 'none', // this leaves the source code as close as possible to the original (when packaging we set this to 'production')

  entry: './src/extension.ts', // the entry point of this extension, 📖 -> https://webpack.js.org/configuration/entry-context/
  output: {
    // the bundle is stored in the 'dist' folder (check package.json), 📖 -> https://webpack.js.org/configuration/output/
    path: path.resolve(__dirname, 'dist'),
    filename: 'extension.js',
    libraryTarget: 'commonjs2'
  },
  externals: {
    vscode: 'commonjs vscode' // the vscode-module is created on-the-fly and must be excluded. Add other modules that cannot be webpack'ed, 📖 -> https://webpack.js.org/configuration/externals/
    // modules added here also need to be added in the .vscodeignore file
  },
  resolve: {
    // support reading TypeScript and JavaScript files, 📖 -> https://github.com/TypeStrong/ts-loader
    extensions: ['.ts', '.js']
  },
  module: {
    rules: [
      {
        test: /\.ts$/,
        exclude: /node_modules/,
        use: [
          {
            loader: 'ts-loader'
          }
        ]
      },
    ]
  },
  devtool: 'nosources-source-map',
  infrastructureLogging: {
    level: "log", // enables logging required for problem matchers
  },
};
module.exports = [ extensionConfig ];

src/assets/main.js
(function() {
    const vscode = acquireVsCodeApi();

    const query = e => document.querySelector(e);
    const queryAll = e => document.querySelectorAll(e);
    const postMsg = (type, data) => vscode.postMessage({ type, data });

    // const oldState = vscode.getState();
    // setTimeout(postMsg('init'), 1000);

    const prefixDom = query('#prefix');
    const suffixDom = query('#suffix');
    const versionTypeDom = query('#versionType');
    const editPkgDom = query('#editPkg');
    const submitBtn = query('#submitBtn');

    const formData = {
        prefix: '',
        suffix: '',
        versionType: 'patch',
        editPkg: true,
    };
    const updateForm = () => postMsg('formChange', formData);

    prefixDom.addEventListener('change', () => {
        formData.prefix = prefixDom.value;
        updateForm();
    });
    suffixDom.addEventListener('input', () => {
        formData.suffix = suffixDom.value;
        updateForm();
    });
    versionTypeDom.addEventListener('change', () => {
        formData.versionType = versionTypeDom.value;
        updateForm();
    });
    editPkgDom.addEventListener('change', () => {
        formData.editPkg = editPkgDom.checked;
        updateForm();
    });
    submitBtn.addEventListener('click', () => {
        postMsg('submit', formData);
    });

    const setPrefixOptions = (data) => {
        const options = data.map((item, index) => {
            return `<option ${index === 0 ? 'selected' : ''} value="${item}">${item}</option>`;
        });
        prefixDom.innerHTML = options.join('');
        formData.prefix = data[0] || '';
        updateForm();
    };
    const updateTag = (tag) => {
        const tagDom = query('#tag');
        tagDom.innerHTML = tag;
    };
    
    window.addEventListener('message', event => {
        const msg = event.data;
        switch (msg.type) {
            case 'prefixOptions':
                setPrefixOptions(msg.data);
                break;
            case 'updateTag':
                updateTag(msg.data);
                break;
            case 'disableSubmit':
                submitBtn.disabled = msg.data;
                break;
        }
    });
})();


src/provider.ts
import * as vscode from 'vscode';
import { Repository } from './type';
import { format } from 'date-fns';

const path = require('path');
const { Uri } = vscode;
import { log, withProgress } from './utils';

let getGitTimes = 0;
export default class TagProvider implements vscode.WebviewViewProvider {

    public static readonly viewType: string = 'gen.tags';
    private _view?: vscode.WebviewView;
    public repo?: Repository;
    public tags: string[] = [];
    public formData: {
        prefix: string;
        suffix: string;
        versionType: string;
        editPkg: boolean;
    } = {
        prefix: '',
        suffix: '',
        versionType: 'patch',
        editPkg: true,
    };
    public newTag: string = '';
    private initLoading: any;
    private pkgContent: any = {};
    
    constructor(private readonly _extensionUri: vscode.Uri) {}

    public resolveWebviewView(
        webviewView: vscode.WebviewView,
        context: vscode.WebviewViewResolveContext,
        _token: vscode.CancellationToken
    ) {
        this._view = webviewView;
        webviewView.webview.options = {
            // Allow scripts in the webview
			enableScripts: true,
            localResourceRoots: [
				this._extensionUri
			]
        };
        webviewView.webview.html = this._getHtmlForWebview(webviewView.webview);
        webviewView.webview.onDidReceiveMessage(msg => {
			switch (msg.type) {
                case 'init': 
                    this.init();
                    break;
                case 'formChange':
                    this.formData = msg.data;
                    this.genTag();
                    break;
                case 'submit':
                    this.handleSubmit();
                    break;
			}
		});
    }

    private _getHtmlForWebview(webview: vscode.Webview) {
        const scriptUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'src', 'assets', 'main.js'));
        const styleVSCodeUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'src', 'assets', 'vscode.css'));
        const styleResetUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'src', 'assets', 'reset.css'));
		const styleMainUri = webview.asWebviewUri(vscode.Uri.joinPath(this._extensionUri, 'src', 'assets', 'main.css'));

        const nonce = this.getNonce();

        return `<!DOCTYPE html>
        <html lang="en">
        <head>
            <meta charset="UTF-8">
            <!--
                Use a content security policy to only allow loading images from https or from our extension directory,
                and only allow scripts that have a specific nonce.
            -->
            <meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src ${webview.cspSource}; script-src 'nonce-${nonce}';">
            <meta name="viewport" content="width=device-width, initial-scale=1.0">
            <link href="${styleResetUri}" rel="stylesheet">
            <link href="${styleVSCodeUri}" rel="stylesheet">
            <link href="${styleMainUri}" rel="stylesheet">
            <title>Cat Colors ${scriptUri}</title>
        </head>
        <body>
            <div class="tags-wrap">
                <div id="repo-list"></div>
                <div class="tags-form">
                    <p class="tags-label">Tag 前缀</p>
                    <select id="prefix" class="tags-target-branch branches-select form" name="tag_prefix">
                    </select>
                    <p class="tags-label">版本类型</p>
                    <select id="versionType" class="tags-target-branch branches-select form" name="version_type">
                        <option value="major">major</option>
                        <option value="minor">minor</option>
                        <option value="patch" selected>patch</option>
                        <option value="RC">RC</option>
                    </select>
                    <p class="tags-label">Tag 后缀</p>
                    <input id="suffix" class="tags-title form" type="text" name="tag_suffix" />
                    <div class="tags-checkbox">
                        <input id="editPkg" class="checkbox" checked type="checkbox" name="edit_pkg">
                        <label for="editPkg">将生成的 tag 存到 package.json#tag 中，并提交更改。</label>
                    </div>
                    <p class="tag-value" id="tag"></p>

                    <button class="tags-btn" id="submitBtn">推送 Tag</button>
                </div>
            </div>
            <script nonce="${nonce}" src="${scriptUri}"></script>
        </body>
        </html>`;
    }

    async getGitInfo() {
        const gitExtension = vscode.extensions.getExtension('vscode.git')?.exports;
        // if (!gitExtension.isActive) {
        //     await gitExtension.activate?.();
        // }
        const git = gitExtension?.getAPI(1);
        const repo: Repository = git?.repositories[0];
        if (repo) {
            this.repo = repo;
        }
    }

    async init() {
        this.disableSubmit();
        if (this.initLoading) {
            this.initLoading.res();
            this.initLoading = null;
        }
        this.initLoading = await withProgress('Gen Tags: 初始化...');
        const pkgStr = await this.optFile('./package.json') || '';
        try {
            this.pkgContent = JSON.parse(pkgStr);
            const prefix = this.pkgContent['tagPrefix'] || [];
            this.postMsg('prefixOptions', prefix);
        } catch (e) {
            log('package.json 解析失败');
        }

        const fn = (res: any) => {
            getGitTimes++;
            if (getGitTimes > 5) {
                this.initLoading.res();
                log('获取 git 信息失败');
                res();
            } else {
                this.getGitInfo();
                if (!this.repo) {
                    setTimeout(() => fn(res), 1000);
                } else {
                    res();
                }
            }
        };
        await new Promise(res => fn(res));

        if (!this.repo) {
            this.initLoading.res();
            return;
        }
        await this.repo.fetch().then(() => {
            this.repo?.getRefs({}).then(res => {
                this.tags = res
                    // type: 2 表示 tag
                    .filter(item => item.type === 2 && item.name !== undefined)
                    .map(item => item.name as string);
                    this.initLoading.res();
                this.genTag();
            });
        }).catch(() => {
            this.initLoading.res();
            log('获取 tags 失败');
        });
        this.disableSubmit(false);
    }

    postMsg(type: string, data: any) {
        this._view?.webview.postMessage({ type, data });
    }
    disableSubmit(disable = true) {
        this.postMsg('disableSubmit', disable);
    }

    getWorkspaceFilePath(fileName: string) {
        const workspaceFolders = vscode.workspace.workspaceFolders;
        if (workspaceFolders && workspaceFolders.length > 0) {
            const workspaceRoot = workspaceFolders[0].uri.fsPath;
            return path.join(workspaceRoot, fileName);
        } else {
            vscode.window.showInformationMessage('No workspace folder is open.');
            return '';
        }
    }
    
    async optFile(filePath: string, content?: string) {
        const fullPath = this.getWorkspaceFilePath(filePath);
        const fileUri = Uri.file(fullPath);
        try {
            if (content) {
                await vscode.workspace.fs.writeFile(fileUri, Buffer.from(content));
            } else {
                const byteArray = await vscode.workspace.fs.readFile(fileUri);
                const fileContent = Buffer.from(byteArray).toString('utf8');
                return fileContent;
            }
        } catch (err) {
            log(`${filePath}: ${err}`);
            throw err;
        }
    }
    
    resolveTags(tags: string[]) {
        return tags.map(item => {
            const match = item.match(/\d+\.\d+\.\d+/);
            return match ? match[0] : '';
        }).filter(item => item !== '');
    }
    
    latestTag(data: string[]) {
        const tags = this.resolveTags(data);
        const tag = tags.sort((a, b) => {
            const aArr = a.split('.');
            const bArr = b.split('.');
            if (aArr[0] === bArr[0]) {
                if (aArr[1] === bArr[1]) {
                    return Number(aArr[2]) - Number(bArr[2]);
                }
                return Number(aArr[1]) - Number(bArr[1]);
            }
            return Number(aArr[0]) - Number(bArr[0]);
        }).pop();
        return tag || '0.0.0';
    }
    
    genTag() {
        const { prefix, suffix, versionType } = this.formData;
        let version = this.latestTag(this.tags);
        const updateIndex = ['major', 'minor', 'patch'].indexOf(versionType);
        if (updateIndex > -1) {
            const versionArr = version.split('.');
            versionArr[updateIndex] = String(Number(versionArr[updateIndex]) + 1);
            for (let i = updateIndex + 1; i < versionArr.length; i++) {
                versionArr[i] = '0';
            }
            version = versionArr.join('.');
        } else {
            version = version + '-' + versionType + '-' + format(new Date(), 'yyyyMMddHHmmss');
        }
        this.newTag = prefix + version + suffix;
        this.postMsg('updateTag', this.newTag);
    }
    
    async handleSubmit() {
        this.disableSubmit();
        const p = await withProgress('Gen Tags: 推送 Tag...');
        try {
            if (this.formData.editPkg) {
                this.pkgContent.tag = this.newTag;
                this.optFile('./package.json', JSON.stringify(this.pkgContent, null, 4)).catch(() => {
                    throw new Error('package.json#tag 设置失败');
                });
                const fullPath = this.getWorkspaceFilePath('package.json');
                await this.repo?.add([fullPath]).catch(() => {
                    throw new Error('提交 package.json 失败');
                });
                await this.repo?.commit('build: update package.json#tag').catch(() => {
                    throw new Error('提交 package.json 失败');
                });
            }
            await this.repo?.tag(this.newTag, 'HEAD').catch(() => {
                throw new Error('创建 Tag 失败');
            });
            await this.repo?.push('origin', this.newTag).catch(() => {
                throw new Error('推送 Tag 失败');
            });
            p.res();
            this.init();
            this.disableSubmit(false);
        } catch (e) {
            log((e as Error).message || '推送失败');
            p.res();
            this.disableSubmit(false);
        }
    }
    
    getNonce() {
        let text = '';
        const possible = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        for (let i = 0; i < 32; i++) {
            text += possible.charAt(Math.floor(Math.random() * possible.length));
        }
        return text;
    }
}
